File: leetcode/go/leetcode_372.go
// https://leetcode.cn/problems/super-pow/
package leetcode

/*
13/2
(10+3)/2
113/2=26
100/2+10/2+3/2
(50 + 3) / 2 =26

除个位外, 若当前位不能整除, 则 下一位+10
*/
func superPow(a int, b []int) int {
	const mod = 1337
	ans := 1
	a %= mod
	for len(b) != 0 {
		n := len(b)
		for i := 0; i < n; i++ {
			if b[i]%2 == 1 {
				if i == n-1 {
					ans = ans * a % mod
				} else {
					b[i+1] += 10
				}
			}
			b[i] >>= 1
		}
		for len(b) != 0 && b[0] == 0 {
			b = b[1:]
		}
		a = a * a % mod
	}
	return ans
}

File: leetcode/leetcode_372.cpp
/**
 * @file leetcode_372.cpp
 * @author Cattle_Horse (CattleHorse@duck.com)
 * @brief https://leetcode.cn/problems/super-pow/description/
 * @version 0.1
 * @date 2024-06-08
 *
 * @copyright Copyright (c) 2024
 *
 */

#include <vector>
class Solution {
private:
    constexpr static int MOD{1337};
    static int qpow(int a, int b) {
        int ans{1};
        for (a %= MOD; b != 0; b >>= 1, a = a * a % MOD)
            if (b & 1) ans = ans * a % MOD;
        return ans % MOD;
    }

public:
    // a^b = (a^10)^(b/10) * a^(b%10)
    int superPow(int a, std::vector<int>& b) {
        int ans = 1;
        for (int i = b.size() - 1; i >= 0; --i) {
            ans = ans * qpow(a, b[i]) % MOD;
            a = qpow(a, 10);
        }
        return ans;
    }
};